pip/pip-333.md

```diff
@@ -0,0 +1,387 @@
+server.
+- `DefaultSslContextBuilder` - It internally creates the `javax.net.ssl.SSLContext` using the ".crt", ".pem" and ".key"
+files for the web server.
+
+# Motivation
+Apache Pulsar's TLS encryption configuration is not pluggable. It only supports file-based certificates. This makes 
+Pulsar difficult to adopt for organizations that require loading TLS certificates by other mechanisms.
+
+# Goals
+The purpose of this PIP is to introduce the following:
+
+- Provide a mechanism to plugin a custom SSL Factory that can generate SSL Context and SSL Engine. 
+- Simplify the Pulsar code base to universally use `javax.net.ssl.SSLContext` and reduce the amount of code required to
+build and configure the SSL context taking into consideration backwards compatibility.
+
+## In Scope
+
+- Creation of a new abstract class `SSLFactory` that can generate a SSL Context, Client SSL Engine and Server SSL 
+Engine. It should support refreshing the certificates at a configurable interval.
+- Creation of a default implementation of `SSLFactory` that supports loading the SSL Context and SSL Engine via
+file-based certificates. Internally it will use the SecurityUtility and KeyStoreSSLContext.
+- Modify the Pulsar Components to support the `SSLFactory` instead of the SslContextAutoRefreshBuilder, SecurityUtility
+and KeyStoreSSLContext.
+- Remove the SslContextAutoRefreshBuilder and all its implementations.
+
+# High Level Design
+```mermaid
+flowchart TB
+    Proxy.DirectProxyHandler --> Sslfactory
+    Proxy.AdminProxyHandler --> Sslfactory
+    Proxy.ServiceChannelInitializer --> Sslfactory
+    Broker.PulsarChannelInitializer --> Sslfactory
+    Client.PulsarChannelInitializer --> Sslfactory
+    Broker.WebService --> JettySSlContextFactory
+    Proxy.WebServer --> JettySSlContextFactory
+    PulsarAdmin --> AsyncHttpConnector
+    AsyncHttpConnector --> Sslfactory
+    JettySSlContextFactory --> Sslfactory
+    Client.HttpLookup.HttpClient --> Sslfactory
+    Sslfactory -.-> DefaultSslFactory
+    Sslfactory -.-> CustomSslFactory
+```
+
+# Detailed Design
+
+## Design and Implementation Details
+
+### Pulsar Common Changes
+
+A new abstract class called `SslFactory` that provides public methods to create a SSL Context, Client SSL Engine and
+Server SSL Engine. The SSL Context class returned will be of type `javax.net.ssl.SSLContext`. This class will refresh 
+the SSL Context at the configured interval. This refresh behavior is the same as the 
+current `SslContextAutoRefreshBuilder`.
+
+```java
+public abstract class SslFactory {
+    public SslFactory(long certRefreshInSec);
+    public abstract SSLEngine getClientSslEngine(String peerHost, int peerPort);
+    public abstract SSLEngine getServerSslEngine() throws GeneralSecurityException, IOException;
+    public abstract boolean needsUpdate();
+    public abstract void update() throws Exception;
```

Comment:
Add javadocs for `needsUpdate` and `update` to make it clear how they are are intended to work.

Reply:
Makes sense. Added it.